// @parcel/cache/lib/LMDBCache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LMDBCache = void 0;
function _crypto() {
  const data = _interopRequireDefault(require("crypto"));
  _crypto = function () {
    return data;
  };
  return data;
}
function _stream() {
  const data = _interopRequireDefault(require("stream"));
  _stream = function () {
    return data;
  };
  return data;
}
function _path() {
  const data = _interopRequireDefault(require("path"));
  _path = function () {
    return data;
  };
  return data;
}
function _util() {
  const data = require("util");
  _util = function () {
    return data;
  };
  return data;
}
function _core() {
  const data = require("@parcel/core");
  _core = function () {
    return data;
  };
  return data;
}
function _fs() {
  const data = require("@parcel/fs");
  _fs = function () {
    return data;
  };
  return data;
}
var _package = _interopRequireDefault(require("../package.json"));
function _lmdb() {
  const data = _interopRequireDefault(require("lmdb"));
  _lmdb = function () {
    return data;
  };
  return data;
}
var _FSCache = require("./FSCache");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// flowlint-next-line untyped-import:off
const pipeline = (0, _util().promisify)(_stream().default.pipeline);

/**
 * See `LMDBCache::setLargeBlob`
 */

class LMDBCache {
  // $FlowFixMe

  constructor(cacheDir) {
    this.fs = new (_fs().NodeFS)();
    this.dir = cacheDir;
    this.fsCache = new _FSCache.FSCache(this.fs, cacheDir);
    this.store = _lmdb().default.open(cacheDir, {
      name: 'parcel-cache',
      encoding: 'binary',
      compression: true
    });
  }
  ensure() {
    return Promise.resolve();
  }
  serialize() {
    return {
      dir: this.dir
    };
  }
  static deserialize(opts) {
    return new LMDBCache(opts.dir);
  }
  has(key) {
    return Promise.resolve(this.store.get(key) != null);
  }
  get(key) {
    let data = this.store.get(key);
    if (data == null) {
      return Promise.resolve(null);
    }
    return Promise.resolve((0, _core().deserialize)(data));
  }
  async set(key, value) {
    await this.setBlob(key, (0, _core().serialize)(value));
  }
  getStream(key) {
    return this.fs.createReadStream(_path().default.join(this.dir, key));
  }
  setStream(key, stream) {
    return pipeline(stream, this.fs.createWriteStream(_path().default.join(this.dir, key)));
  }
  getBlob(key) {
    let buffer = this.store.get(key);
    return buffer != null ? Promise.resolve(buffer) : Promise.reject(new Error(`Key ${key} not found in cache`));
  }
  async setBlob(key, contents) {
    await this.store.put(key, contents);
  }
  getBuffer(key) {
    return Promise.resolve(this.store.get(key));
  }
  #getFilePath(key, index) {
    return _path().default.join(this.dir, `${key}-${index}`);
  }
  async hasLargeBlob(key) {
    if (!(await this.has(key))) {
      return false;
    }
    const entry = await this.get(key);
    if ((entry === null || entry === void 0 ? void 0 : entry.type) !== 'LARGE_BLOB') {
      return false;
    }
    return this.fsCache.hasLargeBlob(entry.largeBlobKey);
  }
  async getLargeBlob(key) {
    if (!(await this.has(key))) {
      throw new Error(`No large blob entry found for key=${key}`);
    }
    const entry = await this.get(key);
    if ((entry === null || entry === void 0 ? void 0 : entry.type) !== 'LARGE_BLOB') {
      throw new Error(`Invalid entry at large blob key=${key}`);
    }
    return this.fsCache.getLargeBlob(entry.largeBlobKey);
  }

  /**
   * Set large blob into LMDB.
   * This stores large blobs as files on a delegate FSCache,
   * but uses an LMDB entry to provide transactional behaviour.
   *
   * On its own the FSCache implementation is not transactional and
   * may result in corrupted caches. Furthermore, it may result in
   * partially written or read caches, where we are concatenating bytes
   * from different cache writes.
   */
  async setLargeBlob(key, contents, options) {
    // $FlowFixMe flow libs are outdated but we only support node>16 so randomUUID is present
    const largeBlobKey = `${key}_${_crypto().default.randomUUID()}`;
    await this.fsCache.setLargeBlob(largeBlobKey, contents, options);
    await this.set(key, {
      type: 'LARGE_BLOB',
      largeBlobKey
    });
  }
  async deleteLargeBlob(key) {
    if (!(await this.has(key))) {
      return;
    }
    const entry = await this.get(key);
    if ((entry === null || entry === void 0 ? void 0 : entry.type) !== 'LARGE_BLOB') {
      return;
    }
    await this.store.remove(key);
    return this.fsCache.deleteLargeBlob(entry.largeBlobKey);
  }
  refresh() {
    // Reset the read transaction for the store. This guarantees that
    // the next read will see the latest changes to the store.
    // Useful in scenarios where reads and writes are multi-threaded.
    // See https://github.com/kriszyp/lmdb-js#resetreadtxn-void
    this.store.resetReadTxn();
  }
}
exports.LMDBCache = LMDBCache;
(0, _core().registerSerializableClass)(`${_package.default.version}:LMDBCache`, LMDBCache);
